src/Orleans.Runtime/Core/InsideRuntimeClient.cs

```diff
@@ -812,15 +812,26 @@ public void RemoveExtension(IGrainExtension handler)
             return false;
         }
 
+        private ExtensionInvoker GetCurrentExtensionInvoker()
+        {
+            return (RuntimeContext.CurrentActivationContext.ContextType == SchedulingContextType.SystemTarget)
+                ? (RuntimeContext.CurrentActivationContext as SchedulingContext)?.SystemTarget.ExtensionInvoker
+                : GetCurrentActivationData().ExtensionInvoker;
+        }
+
         private ActivationData GetCurrentActivationData()
         {
-            var activationData = (IActivationData) (RuntimeContext.CurrentActivationContext as SchedulingContext)?.Activation;
-            if (activationData == null)
-                throw new InvalidOperationException("Attempting to GetCurrentActivationData when not in an activation scope");
-            return (ActivationData)activationData;
+            if (TryGetCurrentActivationData(out ActivationData activationData)) return activationData;
+            throw new InvalidOperationException("Attempting to GetCurrentActivationData when not in an activation scope");
```

Comment:
Prefer we refactor this throw into a separate local function inside of `GetCurrentActivationData` so that the method can be inlined